### PR TITLE
Show Tibet as part of China on map

### LINKED ISSUE #6006: Remove commit 4e47c8 (Remove Tibet from the map)
Please remove commit 4e47c8aba3bf146292f6e9d38aa0151c383e6c33.

This commit deeply hurt the Chinese Piwik community and should be immediately removed from Piwik.

### PR DESCRIPTION
**Pull request by sgiehl**
This Pull Request is not yet complete.

There might be still some places where old data might be shown as "Unknown" now.

And as old visits won't be changed, segments for showing visits from Tibet might have incorrect results, need to check if there is an easy way to handle that.

fixes #6006

### CODE REVIEW

**Comment by sgiehl** [comment]
@Findus23  We need to remove the flag for Tibet as soon as this gets merged in order to get the tests fixed

**Comment by gameblabla** [comment]
I hope this doesn't get merged.

**Comment by phodal** [comment]
@gameblabla Hi, had your used a map? such as Google map, Bing map or Apple map, the last commit about "free tibet" was a joke!

**Comment by gameblabla** [comment]
@phodal, i know them but i do not get your point. 
Perhaps it could done another way, keep tibet as a country but mark it as a "disputed territory".

**Comment by fxzjshm** [comment]
@gameblabla Tibet belongs to China, and there is no dispute

**Comment by gameblabla** [comment]
@fxzjshm There's no need to be hostile, i'm trying to find a good compromise. There are people who agree with you and there are people who don't agree. I believe piwik should not be one-sided.

@sgiehl, forgive me but is my idea okay with you ? This might help the "Unknown" stuff without breaking things. (i admit i know very little php)

**Comment by phodal** [comment]
@gameblabla, did any country recognized Tibet was a country ? America? England ? France? If so, could you list the country name ?

May be for now, India thought Tibet was a country? Because a lot of Indian army was within Chinese  borders. Are you come from India, just prepare discuss the disputed territory side of China?

I remember in 1906, [Convention Between Great Britain and China Respecting Tibet](https://en.wikisource.org/wiki/Convention_Between_Great_Britain_and_China_Respecting_Tibet）The Government of China also undertakes not to permit any other foreign State to interfere with the territory or internal administration of Tibet.

So please don't permit interfere with the territory or internal administration of Tibet.

**Comment by gameblabla** [comment]
I'm not sure if this is the place to discuss about this but i will anyway.
@phobal There's India which has troops in Ladakh, which is close to the tibetan border. Britain also recognized it as an autonomous region (as the "administration of Tibet" in the Convention of 1906) until 2008. Before Nixon, the US also recognized Tibet. To this day, it is still disputed by the Dalai-Lama.
So I think it is enough proof that the territory, while indeed under China's control, is still in dispute.
Hence my idea

**Comment by fengkaijia** [comment]
@gameblabla There's no point to discuss historical claims on this. The Taiwan Authority of China claims Outer-Mongolia until 2002, then should we prepare an addition map that has Mongolia inside China? If we don't build map base on the current situation, then I'm sure we need to build hundreds of maps.

**Comment by gameblabla** [comment]
@fengkaijia, the difference is that no one claims the territory for the "Mongol Empire" or "the empire that sun never sets" whereas it's different with Tibet, even now. That's not counting Australia welcoming tibetan refugees.

Ultimately, it's up to piwik's devs to decide if they should go ahead with the merge or not.

**Comment by fengkaijia** [comment]
@gameblabla No no you are mixing up with the refugee concept. Russian welcomed Edward Snowden, it doesn't mean Russia claimed the US territory. Tibet is a part of China, that's what Australia [claims](http://pmtranscripts.pmc.gov.au/release/transcript-16883):

> The Australian side confirmed the position of successive Australian Governments since 1972 that Australia respects China's sovereignty and territorial integrity, including in relation to Tibet and Xinjiang.

**Comment by gameblabla** [comment]
@fengkaijia, yes you're right, i wasn't denying that. But still, i was pointing out that some people (including tibetans in australia) still do not agree with China on this.

I'll let the devs decide on this.

**Comment by sgiehl** [comment]
There is only one reason why we will change this. We want to show a map that matches other maps of analytics tools like Google Analytics and others. So there won't be any reason in the future to discuss stuff like that. Sure, there might still be some stuff incorrect, but this one can be changed by reverting an old commit, while other changes aren't easily possible without #11929

**Comment by pengsamkee** [comment]
hey，fuck your mum

**Comment by gameblabla** [comment]
I understand the technical reasons. 
Still, it is not desirable and having this PR being merged would be disappointing.
The whole point of Piwik was to not become like Google Analytics, if it did,
then what is the point ?

If this PR should be merged, then most likely you'll get more issues about Taiwan and (eventually) Hong Kong.
Do you want to get even more pressure ? Sometimes, it is best not to cooperate to avoid further harassment.

I do wonder what @mattab think of this? He was the one who made the original commit after all.

**Comment by sgiehl** [comment]
@gameblabla  I didn't mean Piwik wants to become like Google Analytics.
And I don't want to continue any discussion about that. We aim to use official maps and this is a first step. So if official maps will change we aim to change our maps as well.
And if https://github.com/piwik/piwik-map-generator/pull/1 will help to fix our map generator and we generate some new maps, Tibet would be part of China again as well.

**Comment by gameblabla** [comment]
Fine then, i will not insist further. I am considering a fork.
Have a nice day

**Comment by sgiehl** [comment]
Remaining failing tests should be fixed as soon as flag was removed from piwik/icons

**Comment by fengkaijia** [comment]
@sgiehl I just sent a PR piwik/piwik-icons#14 to remove the flag.

**Comment by fengkaijia** [comment]
Plus, I wonder if the UserCounty is applied for old visit, [plugins/UserCountry/Visitor.php](plugins/UserCountry/Visitor.php) shall also be changed.

**Comment by Findus23** [comment]
@fengkaijia's pull request is merged and reflected in https://github.com/piwik/piwik/pull/11904

**Comment by WG-Com** [comment]
@gameblabla NO Psycholagny!

**Comment by Findus23** [comment]
@WG-Com 

I'd seriously recommend you to stop being rude to other people no matter if you agree with their opinion or you don't.

**Comment by WG-Com** [comment]
@Findus23 i agree with you , but you must stop that people stop attack other country

**Comment by Findus23** [comment]
I don't think @gameblabla did issue military threats over some disagreement over a small feature of a software on the internet.

In a normal discussion every position should be heard as long as it isn't rude, overly emotional or in caps lock.

**Comment by WG-Com** [comment]
@Findus23   you know freedom has bottom line right ?

**Comment by phodal** [comment]
@WG-Com just like you said, stop attack other country, please.

**Comment by Findus23** [comment]
@WG-Com How does calling for an objective discussion make me a nazi? And what does my comment have to do with Austria or Germany?

You are calling for respect, but for me this means accepting that other people in other countries have different opinions and together finding a compromise everyone can live with. 

If you disagree with having open discussions where everyone can state their opinion, I beg you to stay away from piwik and even other open source projects as personal attacks will only delay the solution of this issue.

**Comment by mattab** [comment]
> I do wonder what @mattab think of this? He was the one who made the original commit after all.

Putting Tibet map was a small act of raising awareness about the military invasion of Tibet by China in 1950, and the subsequent destruction of Tibetan culture and freedom of thought by China. China occupation has resulted in the deaths of hundreds of thousands of Tibetans and the imprisonment and torture of thousands more. Plus the destruction of shrines, monasteries... 

For most Chinese people, this is actually a good thing as according to them Tibet needed modernization, alphabetisation, etc. and that part was true. But it is really bad how China invaded Tibet and didn't leave these people to sort out their modernization needs on their own, rather than force it upon them. China could have helped Tibet to become a great nation with funding and kindness and respect rather than invade and kill them. And we know that Tibet is a huge natural resource for China (fresh water, mineral). As always, it's not so much about the "great china nation" as it really is about the natural resources. 

For reference, similar invasions happened recently in western countries such as the USA when they destroyed the native American culture, or in Australia when they destroyed the aboriginal culture. It could and should have been done differently in an ideal world. But we don't live in this ideal world yet and China did indeed successfully invade Tibet with their army, and killed 1+ million people there. 

I think that Chinese people are blinded by the extremely potent propaganda by their leaders, school system, etc. It's interesting how easy it is to influence human beings in believing what you want them to believe, if you control the school system and media. Obviously in our western society we are also blinded by propaganda about consumerism, capitalism, nationalism, and all the rest. It's not just China who are blinded by their leaders' propaganda.

I like the idea of having put the huge nation of Tibet on the map of Piwik Analytics for the last 4 years. But now that some other team members are pushing for restoring the "standard world map", it is hard for me alone to resist this push by my other colleagues.

**Comment by WG-Com** [comment]
@Findus23 i am not saying you are a Nazi ,  i am asking you what feeling when i say that, now some people is saying that to us , it is attack , just watch @mattab comments , all knowledge not truth

**Comment by WG-Com** [comment]
@phodal  “I like the idea of having put the huge nation of Tibet on the map of Piwik Analytics for the last 4 years. But now that some other team members are pushing for restoring the "standard world map", it is hard for me alone to resist this push by my other colleagues.”

你够了

**Comment by gameblabla** [comment]
Thank you @mattab for chiming in. I've honestly lost all hope too.
Trying to reason them by portraying China as Human Rights infringers is like yelling at a cloud :
it goes right through them, not matter your reasoning.
The middle-class in China do not even care for the poor condition of their own workers at Foxconn,
why would they care for the slaughter for tibetans 50 years ago or the persecution of Falul-Gong 18 years ago ?

@sgiehl seemingly wants to forget all of this happened and wants to merge this as quick as possible,
despite the pointless trolling he got and some chinese ppl went even as far as to report him on github.
That's not even mentioning how some users here are abusing the dislike button to try to silence critics.

I was foolish to think that i could change things, the code itself spoke.
I think from now on i will just block Piwik until this changes.

**Comment by jijiechen** [comment]
@mattab Thank you for sharing your thoughts. Although your opinion is hard to be agreed according to the reality of current status of Tibet.
You mentioned the modernization of Tibet, but I'd like to remind you that, Tibet was never a country before 1950, and it won't be one in the future either. The governance in Tibet from Chinese central government was started hundreds of years ago.
So the idea of "building Tibet as a modern country" is just a wish of yourself, and maybe others who share the same unrealistic dream.

**Comment by WG-Com** [comment]
@gameblabla "The middle-class in China do not even care for the poor condition of their own workers at Foxconn so.."

you don't understand why Chinese government and people "do not care" this ，i support this ,because if they Make strict laws，those workers will have no job, I'm serious，Your idea is good, but it is not easy to put it into practice, i promise this is not you think , i know many Foxconn workers , more jobs and salary make more  better , we are not rich,  but better and better every year, all people feel that

**Comment by WG-Com** [comment]
@gameblabla you should update your knowledge about us

**Comment by WG-Com** [comment]
@mattab a book has many chapters , why you just read the last ?  tibet become a part of China in AD 1300-1400, by Mongolian, in 1900-1950 they never become a independent by  Referendum，PLA just enter our own territory, when you say china into tibet 1950s , but you do not know what happend in 1949-1956

**Comment by WG-Com** [comment]
your lack the basic knowledge of tibet , read more

**Comment by happydoubi** [comment]
It's ridiculous that someone, who was and is brainwashed by their prejudiced and hostile propagation, jumps out and says Tibet was invaded. Do you think you can invent new history? 1+ million Tibetan were killed? Haha, can you provide the source of your figure? Do you know the population of Tibetan now? 3.2+ million, let me tell you. You mean in the 1950s, more than 1/3 Tibetan were killed? I know you arrogant westerners think China is a devil. But that's not the truth. Tibetan now live in a happier life and have freedom of religion. Before 1950s, they were still in a slave society, even no personal freedom. You poor and arrogant westerners are those who need to open you eyes. Come to China to see, not stay in your bed to imagine.

**Comment by happydoubi** [comment]
@WG-Com 不用理这些智障了。你知道有些国外傻逼被他们的宣传机构洗脑了，多说无益。估计那俩瞎BB的就是阿三吧，也就打打嘴炮，PLA陈兵边境，很快就有阿三的好看了。

**Comment by bsssdnnh** [comment]
@mattab Again you are a great history inventor. This is the first I've heard that a normal regime change in a civil war is an action of "annexation" or "invasion", HILARIOUS!

You always mention 1950, however, according to "Provisional Constitution of the Republic of China" in 1912, Tibet had officially belonged to Republic of China (now its pathetic government is merely ruling Taiwan and some surround islands though) since then. Moreover, ROC inherited the territory Qing Empire which included Tibet -- I mean, such Part-Whole relation between Tibet and China is much earlier than 1912, let alone your story of 1950. In 1950 it was merely a regime change during a civil war (People's Republic of China vs Republic of China). Here is an example which you are familiar with (if you have sufficient education with a qualified high school diploma)  is here: US Civil War, "United States of America defeated Confederate States of America".

You said "now that some other team members are pushing for restoring the 'standard world map', it is hard for me alone to resist this push by my other colleagues.", oh boy, that's the beauty of DEMOCRACY! That's the greatest thing in the world, right?

@mattab Again, please keep in mind that this is Github and this is Piwik pull request panel, everything should be in pure tech, and never initialize any unnecessary discussion about politics. It is your childish action that almost ruined this project. Piwik is now going business as far as I know, and if something goes business, the associating team should be responsible. All members in such team should be mature and comply to a lot of things even they are against individual opinions. See how many troubles you caused for your team, especially @sgiehl who tries to focus on pure tech. Shame on you!

A final say to @gameblabla , before you accuse Chinese people of so-called "brainwash", check yourself first. If you really want your voice heard, stop being a Keyboardman ("man" for Spiderman and Batman) on virtual world. Come to real world and here is a perfect physical address to whine:
United Nations Headquarters
First Avenue between 42nd and 48th Streets
New York, NY, 10017 USA

**Comment by PaniniGelato** [comment]
There are many fact like 'Invadation Tibet' in China History

> Ancient Qin empire conquered six countries including Han, Zhao, Wei, Chu, Yan and Qi. Well, I used to be a citizen of Zhao, by the way. https://en.wikipedia.org/wiki/Qin%27s_wars_of_unification
> 
> Ancient Han empire invaded Huns. https://en.wikipedia.org/wiki/Han%E2%80%93Xiongnu_War
> 
> Ancient Tang empire destroyed Xueyantuo, Tujue, Huihu, etc.  https://en.wikipedia.org/wiki/Emperor_Taizong%27s_campaign_against_the_Western_Regions
> 
> Qing empire merged Xinjiang, Monglia and Tibet into China. 
> https://en.wikipedia.org/wiki/Revolt_of_the_Altishahr_Khojas
> https://en.wikipedia.org/wiki/Mongolia_under_Qing_rule
> https://en.wikipedia.org/wiki/Dzungar%E2%80%93Qing_Wars

There are many descendants of Zhao and other ancient countries/kingdoms in China. Please do try to free us all if you can. Please save us ASAP.

If u can‘t save us, just shut up and leave this repo pure tech. 
@mattab @gameblabla

**Comment by recq-cse** [comment]
**gameblabla mattab**  wish you a mmp!

**Comment by wdhwg001** [comment]
Piwik Issues page is not the right place to talk about political issues. Threatening, spamming and being rude are not the right way to discuss.

But I do believe that following the most countries and the UN decision about country existence and country borders is the right way to do.

I know, we hackers are making free software for mankind.

However, it would be a huge work for Piwik to make maps for all political claims, so making the most people happy is the fair and acceptable option. Liberty is not hurting more people to make less people feel free, right?

Then I believe the story about mistaken position Piwik happened to stand on for years needs to come to an end. And I sencerely wish that this hurting story won't happen in any open-source software project, no matter which country those people come from.

Thanks for the fix, for real.

**Comment by f2ex** [comment]
@gameblabla China's things do not need you to intervene ,  you can pay more attention to your own country's problems .

**Comment by fengkaijia** [comment]
Can someone merge this? It's been like two weeks and no problem has been reported.

**Comment by fengkaijia** [comment]
@sgiehl Looks like you need to regenerate the expected screenshots. The merge of #11579 changes the UI. And any updates when will this got merged? Looks like it's gonna miss the 3.1 window.

**Comment by fengkaijia** [comment]
Sent a PR #12016 removing some leftovers. Guess it still cannot pass the I&T, the error with translation ```Piwik\Plugins\LanguagesManager\TranslationWriter\Filter\EncodedEntities changed``` looks like something outside updates in this branch.

File: plugins/UserCountry/API.php
Comment (tsteur) [inline plugins/UserCountry/API.php]: As those `GroupBy` filters are quite slow I wonder if we could maybe remove this again in Piwik 4 and somehow mention this in an issue or so? Or maybe we could have first a simple check whether there is actually any row containing this label? Like `if($dataTable->getRowFromLabel('1|ti')) { ... groupby filter...} `  

Same could be done in countries. This would not check lowercase but don't know if we actually have it in different cases in the DB?

File: plugins/UserCountry/functions.php
Comment (tsteur) [inline plugins/UserCountry/functions.php]: Just wondering if this is still needed after moving `ti` to `cn` ?

Comment (tsteur) [reply plugins/UserCountry/functions.php]: I see... it might be for visitor log or so

File: plugins/UserCountryMap/javascripts/visitor-map.js
Comment (tsteur) [inline plugins/UserCountryMap/javascripts/visitor-map.js]: Map seems to still work 👍

**Review by tsteur — CHANGES_REQUESTED** [review]
I'd only try to improve the GroupBy to do it only when needed and when the table actually contains such a label, will improve performance a bit as groupBy can take a bit of time (in this case it wouldn't take too long as it would not actually group many, but be still quite a bit faster and take less memory etc). Otherwise good to merge.

File: plugins/UserCountry/API.php
Comment (sgiehl) [reply plugins/UserCountry/API.php]: Checking if an entry exists before adding the group by filters should work, will test that.
Removing in Piwik 4 might not be that easy. We could add an update script to update all entries in log_visit table (which I would do in next major release only as it might run a bit longer), so we can remove some of the checks for visitor log & profile, but the group by would still be required if the archived reports aren't reprocessed.

Comment (tsteur) [reply plugins/UserCountry/API.php]: It is fine to leave it in there as long as we do the check if table has the row at all.

**Review by tsteur — APPROVED** [review]

**Comment by sgiehl** [comment]
Last changes weren't compatible with Datatable Maps. Just pushed an update. Let's see if all tests pass now and then we can finally merge

**Comment by tsteur** [comment]
👍

File: plugins/UserCountry/API.php
Comment (fengkaijia) [inline plugins/UserCountry/API.php]: @sgiehl Thanks for the merge. Should the return value be ```14|cn```?

Comment (sgiehl) [reply plugins/UserCountry/API.php]: sure. I'll fix that typo

**Comment by Glisse1** [comment]
Chinese pressure is so big that no one can resist it seems. 

Say they conquer korea and japan today like they did the massacre in Tibet, next year piwik will update the new map with the new "chinese territories" :)

**Pull request merged**
